hyperlink/_url.py

```diff
@@ -1464,3 +1504,329 @@ def remove(self, name):
+        """Automatically encode any reserved characters in *segments* and
+        return a new DecodedURL wrapping the result of
+        :meth:`~hyperlink.URL.child()`.
+        """
+        if not segments:
+            return self
+        new_segs = [_encode_reserved(s) for s in segments]
+        return type(self)(self._url.child(*new_segs))
+
+    def normalize(self, *a, **kw):
+        "Return a new DecodedURL wrapping the result of :meth:`~hyperlink.URL.normalize()`"
+        return type(self)(self._url.normalize(*a, **kw))
+
+    @property
+    def absolute(self):
+        return self._url.absolute
+
+    @property
+    def scheme(self):
+        return self._url.scheme
+
+    @property
+    def host(self):
+        return _decode_host(self._url.host)
+
+    @property
+    def port(self):
+        return self._url.port
+
+    @property
+    def rooted(self):
+        return self._url.rooted
+
+    @property
+    def path(self):
+        try:
+            return self._path
+        except AttributeError:
+            pass
+        self._path = tuple([_percent_decode(_encode_path_part(p),
+                                            raise_subencoding_exc=True)
+                            for p in self._url.path])
+        return self._path
+
+    @property
+    def query(self):
+        try:
+            return self._query
+        except AttributeError:
+            pass
+        _q = [tuple(_percent_decode(_encode_query_part(x),
+                                    raise_subencoding_exc=True)
+                    if x is not None else None
+                    for x in (k, v))
+              for k, v in self._url.query]
+        self._query = tuple(_q)
+        return self._query
+
+    @property
+    def fragment(self):
+        try:
+            return self._fragment
+        except AttributeError:
+            pass
+        frag = self._url.fragment
+        self._fragment = _percent_decode(_encode_fragment_part(frag),
+                                         raise_subencoding_exc=True)
+        return self._fragment
+
+    @property
+    def userinfo(self):
+        try:
+            return self._userinfo
+        except AttributeError:
+            pass
+        self._userinfo = tuple([_percent_decode(_encode_userinfo_part(p),
+                                                raise_subencoding_exc=True)
+                                for p in self._url.userinfo.split(':', 1)])
+        return self._userinfo
+
+    @property
+    def user(self):
+        return self.userinfo[0]
+
+    @property
+    def uses_netloc(self):
+        return self._url.uses_netloc
+
+    def replace(self, scheme=_UNSET, host=_UNSET, path=_UNSET, query=_UNSET,
+                fragment=_UNSET, port=_UNSET, rooted=_UNSET, userinfo=_UNSET,
+                uses_netloc=_UNSET):
+        """While the signature is the same, this `replace()` differs a little from
+        URL.replace. For instance, it accepts userinfo as a tuple, not
```

Comment:
Why is it different?  Presumably it's to make the encoding easier, if not just possible, but it'd be good to see the reason discussed in the docstring.

Reply:
Because what if you have a username with a colon in it, same as everything else in DecodedURL. Updating the docstring.